[12.0][REF] PIM: new module pim_base and upgrade base_custom_attribute

clementmbr opened the pull request:
Refactor creating an oficial PIM application :

- new module **pim_base**
- rename product_custom_attribute into **pim_custom_attribute**
- add the possibility to create attributes related to existing fields (with passing tests)
- allow to display these "native attributes" with the other attributes or to let them at their original location (depending on context `"include_native_attribute"` )
- refactor moving the code overiding the `fields_view_get()` from the model (with Attributes) to the mixin `"attribute.set.owner.mixin"`

By doing so, the model using **base_custom_attribute** only need to :

1. _inherit the mixin "attribute.set.owner.mixin"
2. add a placeholder `<separator name="attributes_placeholder" />` at
the desired location in the model's form view.
3. add a context `{"include_native_attribute": True}` on the action leading to this form view if it is needed to display attributes related to native fields together with the other "custom" attributes.

clementmbr committed:
[ADD] pim_base module

clementmbr committed:
[REF] rename module name 'product_custom_attribute' to 'pim_custom_attribute'

clementmbr committed:
[REF] move pim_custom_attribute in the PIM menu

clementmbr committed:
[REF] remove UoM menu in pim_base

clementmbr committed:
[IMP] create attribute related to existing fields

Add an option "Native" or "Custom" when creating a new
'attribute.attribute' to allow relating it to an existing
'ir.model.fields' field or a new "manual" one.

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -36,18 +36,44 @@ class AttributeAttribute(models.Model):
+
+        if not self.is_custom:
+            if self._context.get("include_native_attribute"):
+                for native_efield in native_efields:
```
I do not think this is necessary, This is too complex !
_build_attribute_field should do only one thing rendering an attribute.
Remove this native_efield stuff, we do not need it.

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -96,18 +120,29 @@ def _build_attribute_field(self, subgroup):
         for attribute in self:
             att_group = attribute.attribute_group_id
             att_group_name = att_group.name.capitalize()
+            if not att_group:
```
Useless the field is required : https://github.com/akretion/odoo-pim/blob/b4310be93be9777d5fa9b66da66d8fdefa5d2358/base_custom_attribute/models/attribute_attribute.py#L265

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -96,18 +120,29 @@ def _build_attribute_field(self, subgroup):
-        """Return a main_group etree element made of sub_groups for each
-        attribute_group."""
-        main_group = etree.Element("group", name="attributes_group", col="4")
+    def _build_attribute_view(self, native_efields):
```
no need of native_efields here

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -127,13 +162,60 @@ def _build_attribute_view(self):
             setup_modifiers(subgroup)
-            attribute._build_attribute_field(subgroup)
-        return main_group
+            attribute.with_context(
```
no need to repass it in the context it's already in the context

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -127,13 +162,60 @@ def _build_attribute_view(self):
+        inverse="_inverse_attribute_nature",
+        default="custom",
+    )
+    is_custom = fields.Boolean(
```
Do not add a "is_custom".
We can simply in the python code do a attribute_nature == 'custom'.
Less code is always better

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -195,6 +276,23 @@ def _build_attribute_view(self):
         "Sequence in Group", help="The attribute's order in his group"
     )
 
+    @api.depends("is_custom")
+    def _compute_attribute_nature(self):
+        for att in self:
+            att.attribute_nature = "custom" if att.is_custom else "native"
+
+    def _inverse_attribute_nature(self):
+        for att in self:
+            att.is_custom = att.attribute_nature == "custom"
+
+    @api.onchange("attribute_nature")
+    def onchange_attribute_nature(self):
+        self.is_custom = self.attribute_nature == "custom"
+
```
remove all of this

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -254,37 +352,38 @@ def button_add_options(self):
-            # changing the fields values.
-            field_obj = self.env["ir.model.fields"]
-            field = field_obj.browse(vals["field_id"])
+        if not vals.get("is_custom"):
```
useless see comment before

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -456,8 +555,8 @@ def write(self, vals):
-        """ Delete the Attribute's related field when deleting an Attribute"""
-        for attribute in self:
+        """ Delete the Attribute's related field when deleting a custom Attribute"""
+        for attribute in [att for att in self if att.is_custom]:
```
we should do something like
```
field_to_removes = self.filtered(lambda s: s.attribute_nature == 'custom').mapped("field_id")
super()..
field_to_removes.unlink()
```

And add a test for this

sebastienbeau commented on base_custom_attribute/models/attribute_set_owner.py:
```diff
@@ -2,18 +2,46 @@
+            view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu,
+        )
+        if view_type == "form":
+            context_include = self._context.get("include_native_attribute")
```
getting and repassing the context is useless context is already passed

sebastienbeau commented on base_custom_attribute/tests/test_build_view.py:
```diff
@@ -93,46 +114,47 @@ def _check_attrset_visiblility(self, attrs, set_ids):
     def test_group_order(self):
-        views = self.env["res.partner"]._build_attribute_view()
-        groups = [g.get("string") for g in views.getchildren()]
+        eview = self.env["res.partner"]._insert_attribute(deepcopy(self.eview))
```
do not change this test.
They test the build_attribute_view.
instead just a one or two test for testing the insert method

sebastienbeau commented on base_custom_attribute/models/attribute_set_owner.py:
```diff
@@ -2,18 +2,46 @@
+            # Attributes
+            eview_attributes = self.with_context(
+                {"include_native_attribute": context_include}
+            )._insert_attribute(eview)
```
pass a string for the "_insert_attribute" method (move the etree.fromstring(result["arch"]) and etree.tostring(eview_attributes, pretty_print=True) into "_insert_attribute"

sebastienbeau commented on base_custom_attribute/tests/test_build_view.py:
```diff
@@ -173,3 +196,45 @@ def test_render_all_field_type(self):
                 self.assertFalse(attr.get("nolabel", False))
+
+    # TEST on NATIVE ATTRIBUTES
+    def test_include_native_attr(self):
```
The test need to be refactored toooooo long.
First split it (never never test 2 cases in one test), one case per test.

For this case test th emethod field_view_get directly. It will avoid all the logic of deepcopy

sebastienbeau commented on base_custom_attribute/tests/test_build_view.py:
```diff
@@ -33,13 +35,21 @@ def setUpClass(cls):
         cls.loader.update_registry((ResPartner,))
+
+        # Add the 'attributes' placeholder in a partner form view
+        cls.eview = etree.fromstring(cls.env.ref("base.view_partner_form")["arch"])
```
instead of hacking the xml here.
Just create a view in Odoo
```
env["ir.ui.view"].create({....})
```

bealdav commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -254,37 +316,36 @@ def button_add_options(self):
+            for key in list(vals.keys()):
+                if key in field_obj.fields_get().keys():
+                    del vals[key]
+            return super(AttributeAttribute, self).create(vals)
```
```suggestion
            return super().create(vals)
```

clementmbr replied:
done!

sebastienbeau commented on base_custom_attribute/models/attribute_attribute.py:
```diff
@@ -128,12 +150,49 @@ def _build_attribute_view(self):
+
+        return main_att_group
+
+    def _insert_attribute(self, arch):
```
la method _insert_attribute, _remove_native_field mettre dans  attribute.mixin

clementmbr committed:
[REF] Add Attr related to native fields in Attr view

...depending on context "include_native_attribute".

Commit include tests and new "Native Attributes" demo records in
pim_custom_attribute module.
It also moves the code overiding the model's fields_view_get from
"product.template" to the "attribute.set.owner.mixin".

By doing so, the model using base_custom_attribute only need to :

- _inherit the mixin "attribute.set.owner.mixin"
- add a placeholder `<separator name="attributes_placeholder" />` at
the desired location in the model's form view.
- add a context `{"include_native_attribute": True}` on the action
leading to this form view if it is needed to display attributes related
to native fields together with the other "custom" attributes.

clementmbr committed:
[REF] rename modules based on attribute_set and pim

New organization :

- attribute_set (former base_custom_attribute)
- product_attribute_set (former pim_custom_attribute but without menus)
- pim (The "PIM application" former pim_base)
- pim_attrubute_set (depends on product_attribute_set and adds menus in
the PIM application)

clementmbr committed:
[FIX] when including native field not present in form

clementmbr committed:
[REF] rename 'attribute_nature' field into 'nature'

And small FIX in _build_attribute_field()

clementmbr committed:
[IMP] Add REAME to the 4 modules

clementmbr committed:
[FIX] propagate readonly attribute for included native attributes

hparfr reviewed (changes requested):
Great work.

Few remarks:
- grant by default the pim manager rights to admin (or put it in another group, see other modules). 
- If I install product_attribute_set alone, I get the following errors when I try to open a product: `It is impossible to add Attributes on "Product" xml view as there is
not one "<separator name="attributes_placeholder" />" in it.`

hparfr commented:
The new feature of native attributes is really cool. :heart:

hparfr commented:
There is one feature which have been dropped. It's the ability to reference existing records with (multi)select. Why ?

clementmbr commented:
@hparfr 
> * grant by default the pim manager rights to admin (or put it in another group, see other modules).

Ok, i'll add this !
> * If I install product_attribute_set alone, I get the following errors when I try to open a product: `It is impossible to add Attributes on "Product" xml view as there is not one "<separator name="attributes_placeholder" />" in it.`

Indeed I should move these two files :
https://github.com/akretion/odoo-pim/blob/12.0-pim-ref-modules-names/pim_attribute_set/views/product.xml
and
https://github.com/akretion/odoo-pim/blob/12.0-pim-ref-modules-names/pim_attribute_set/views/product_category.xml
...from **pim_attribute_set** to **product_attribute_set** in order to display the product Attributes in the product form with **product_attribute_set** installed alone.

And then let in **pim_attribute_set** only the views for the Product Attributes **menus** in the PIM app.
Do you agree @sebastienbeau ?

clementmbr commented:
> There is one feature which have been dropped. It's the ability to reference existing records with (multi)select. Why ?

It didn't disappear, it's just that you need to active the "Advanced Attribute Option" in the user access rights :
![image](https://user-images.githubusercontent.com/31664455/82897264-c1ae6d00-9f2d-11ea-8ca7-081645635161.png)
I mentioned it in the **attribute_set** [README](https://github.com/akretion/odoo-pim/tree/12.0-pim-ref-modules-names/attribute_set#usage), but I can also add this access right automatically to admin for demo/test purpose...

sebastienbeau commented on attribute_set/models/attribute_attribute.py:
```diff
@@ -36,18 +36,23 @@ class AttributeAttribute(models.Model):
     _order = "sequence_group,sequence,name"
 
-    def _get_attrs(self):
+    def _set_attrs(self):
```
_egt_attrs was better because the method return to you the attrs information.
Using _set word have a meaning of doing an action setting something.
Here we do not set anything

sebastienbeau commented on attribute_set/models/attribute_attribute.py:
```diff
@@ -254,37 +271,36 @@ def button_add_options(self):
                         "an existing ir.model.fields (%s)"
                     )
-                    % field.name,
+                    % field_obj.browse(vals["field_id"]).name,
```
serialized is already a native field so it will be removed. So I think it's useless.
In the UI set it in readonly in case of native

clementmbr replied:
In fact, the `serialized` boolean is already in an [invisible group](https://github.com/akretion/odoo-pim/blob/12.0-pim-ref-modules-names/attribute_set/views/attribute_attribute_view.xml#L27) if we're creating a native attribute.
Should I need to add a readonly condition or that's enough ?

sebastienbeau commented on attribute_set/models/attribute_attribute.py:
```diff
@@ -254,37 +271,36 @@ def button_add_options(self):
+            # before creating the new 'attribute.attribute'
+            for key in list(vals.keys()):
+                if key in field_obj.fields_get().keys():
+                    del vals[key]
```
the code is complex and you call field_get a lot of time

```suggestion
            for key in set(vals).intersection(field_obj._fields):
                del vals[key]
```

sebastienbeau commented on attribute_set/security/ir.model.access.csv:
```diff
@@ -0,0 +1,13 @@
+access_attribute_set_attribute_set_manager,attribute_set_attribute_set,attribute_set.model_attribute_set,base.group_no_one,1,1,1,1
+access_attribute_set_attribute_group_manager,attribute_set_attribute_group,attribute_set.model_attribute_group,base.group_no_one,1,1,1,1
+access_attribute_set_attribute_attribute_manager,attribute_set_attribute_attribute,attribute_set.model_attribute_attribute,base.group_no_one,1,1,1,1
+access_attribute_set_attribute_option_manager,attribute_set_attribute_option,attribute_set.model_attribute_option,base.group_no_one,1,1,1,1
```
please remove giving the following access to group_no_one
```suggestion

```

sebastienbeau commented on attribute_set/tests/test_build_view.py:
```diff
@@ -33,13 +34,31 @@ def setUpClass(cls):
         cls.loader.update_registry((ResPartner,))
+
+        # Create a new inherited view with the 'attributes' placeholder.
+        cls.view = cls.env["ir.ui.view"].create(
```
c'est parfait ça !

sebastienbeau commented on attribute_set/views/attribute_option_view.xml:
```diff
@@ -12,7 +12,7 @@
                         name="value_ref"
                         colspan="2"
-                        groups="base_custom_attribute.group_advanced_attribute"
+                        groups="attribute_set.group_advanced_attribute"
```
do not prefix with the module name. AS we are in the module it will be automatically prefixed before the search

clementmbr replied:
When I try to remove these module prefix on security groups, I got some :
```
WARNING db odoo.addons.base.models.ir_ui_view: The group group_advanced_attribute defined in view attribute.option.form.popup does not exist! 
...
KeyError: ('res.users', <function Users._has_group at 0x7f2742a56840>, 1, 'group_advanced_attribute')
...
  File "/odoo/src/odoo/addons/base/models/res_users.py", line 741, in _has_group
    assert group_ext_id and '.' in group_ext_id, "External ID must be fully qualified"
AssertionError: External ID must be fully qualified
```

So it looks like I have to let it (at least for groups)... is it right?

sebastienbeau commented on pim_attribute_set/views/attribute_attribute.xml:
```diff
@@ -22,7 +22,7 @@
         id="menu_attribute_attribute_action"
-        parent="product_custom_attribute.menu_attribute_in_config_product"
-        sequence="4"
+        parent="pim_attribute_set.main_menu_attribute"
```
pim_attribute_set is not needed as we are in the module pim_attribute_set

sebastienbeau commented on pim_attribute_set/views/attribute_group.xml:
```diff
@@ -5,10 +5,7 @@
-            name="search_view_id"
-            ref="base_custom_attribute.attribute_group_search_view"
-        />
+        <field name="search_view_id" ref="attribute_set.attribute_group_search_view" />
```
same here

sebastienbeau commented on pim_attribute_set/views/attribute_group.xml:
```diff
@@ -18,7 +15,7 @@
         id="menu_attribute_group_action"
-        parent="product_custom_attribute.menu_attribute_in_config_product"
-        sequence="2"
+        parent="pim_attribute_set.main_menu_attribute"
```
same here

sebastienbeau commented on product_attribute_set/tests/test_product_attribute.py:
```diff
@@ -16,7 +16,7 @@ def test_write_attribute_values_text(self):
 
```
I think this test can be remove.
we already test such thing in the base module

clementmbr replied:
In the attribute_set's tests we don't really test to **write** on the created attribute's fields as we do on this one... Should we ?

For the moment we only create attributes and check if they are in the form view with all the correct required and invisible attrs...

sebastienbeau replied:
Move this test in the base module

sebastienbeau commented on product_custom_attribute/migrations/7.0.0.2/post-migration.py:
```diff
@@ -1,21 +0,0 @@
-# Copyright 2013 Akretion (http://www.akretion.com).
```
You can remove this file, it's really old and have no sense now

bealdav commented on pim_attribute_set/views/attribute_set.xml:
```diff
@@ -28,7 +25,7 @@
         id="menu_attribute_set_action"
-        parent="product_custom_attribute.menu_attribute_in_config_product"
-        sequence="1"
+        parent="pim_attribute_set.main_menu_attribute"
```
```suggestion
        parent="main_menu_attribute"
```

bealdav commented on pim/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "PIM",
```
```suggestion
    "name": "Product Informations Management",
```

clementmbr replied:
Yes, with no "s" to "Information" I imagine ;)

bealdav commented on pim/__init__.py:
```diff
@@ -0,0 +1 @@
+# from . import models
```
```suggestion
```

bealdav commented on attribute_set/models/attribute_attribute.py:
```diff
@@ -456,10 +472,10 @@ def write(self, vals):
 
     @api.multi
```
Implicitly deprecated in v12 because invalid in v13
```suggestion
```

clementmbr replied:
Should I remove these `@api.multi` from everywhere in the code ? Or only here to override `unlink()` ?

bealdav replied:
everywhere, yes

bealdav commented on attribute_set/models/attribute_attribute.py:
```diff
@@ -456,10 +472,10 @@ def write(self, vals):
+        fields_to_remove = self.filtered(lambda s: s.nature == "custom").mapped(
+            "field_id"
+        )
+        res = super(AttributeAttribute, self).unlink()
```
```suggestion
        res = super().unlink()
```

bealdav reviewed (changes requested):
Some useless class name in super and misc details

bealdav commented:
> It didn't disappear, it's just that you need to active the "Advanced Attribute Option" in the user access rights :

Not sure about exact name, but suggest to rename in "Advanced Attribute Set" to avoid confusion with standard odoo attribute ?

clementmbr commented:
> > It didn't disappear, it's just that you need to active the "Advanced Attribute Option" in the user access rights :
> 
> Not sure about exact name, but suggest to rename in "Advanced Attribute Set" to avoid confusion with standard odoo attribute ?

Maybe _"Advanced Attribute Set settings"_ if this group is made to receive other "Attribute set configuration" other than just allowing to create attributes pointing to other Odoo objects as it does now.

Otherwise, if we're sure that this group won't include other features, I would suggest a more explicit name like _"Relate Attribute set's attributes to existing models"_

What do you think @bealdav @sebastienbeau ?

bealdav commented:
@clementmbr I let other decide, thanks

clementmbr committed:
[IMP] base.user_admin in Attribute Set and PIM Manager group

clementmbr committed:
[IMP] Move Product's attributes_placeholder to product_attribute_set

clementmbr committed:
[IMP] remove class in super() and other misc improvements

clementmbr committed:
[FIX] remove tests in product_attribute_set

clementmbr committed:
[IMP] remove attribute_set access rights for group_no_one

clementmbr committed:
[REF] remove unecessary module name in External ID

clementmbr commented:
@hparfr @sebastienbeau @bealdav I made most of the changes you asked for in these last commits.

I've still have some questions scattered in the discussion which I might gather here :

- @sebastienbeau should I add a [readonly on attribute's `serialized` field](https://github.com/akretion/odoo-pim/pull/3#discussion_r430443805) when creating a native attribute even if it's already in an invisible group ? 
- @sebastienbeau should I [add a test to write on attributes](https://github.com/akretion/odoo-pim/pull/3#discussion_r430491264) in attribute_set as I deleted these tests in product_attribute_set ?
- @bealdav should I [remove the `api.multi` from the whole repo](https://github.com/akretion/odoo-pim/pull/3#discussion_r431191859) or it was just to be removed here for the unlink() ?
- @sebastienbeau @hparfr should I [rename the attribute_set group allowing to relate attributes to other models](https://github.com/akretion/odoo-pim/pull/3#issuecomment-634716897) _"Relate Attribute set's attributes to existing models"_ instead of _"Advanced Attribute Option"_ ?

Thank you very much ! :sparkles:

bealdav reviewed (approved):
LGTM, no test

rvalyi commented:
> @hparfr @sebastienbeau @bealdav I made most of the changes you asked for in these last commits.
> 
> I've still have some questions scattered in the discussion which I might gather here :
> 
> * @sebastienbeau should I add a [readonly on attribute's `serialized` field](https://github.com/akretion/odoo-pim/pull/3#discussion_r430443805) when creating a native attribute even if it's already in an invisible group ?

I would say no need to.

> * @bealdav should I [remove the `api.multi` from the whole repo](https://github.com/akretion/odoo-pim/pull/3#discussion_r431191859) or it was just to be removed here for the unlink() ?

I would say everywhere it's deprecated (it's the new default).
In fact it seems it is tolerated in the OCA in 12.0:
https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0
If you browse an emblematic OCA module you would find them:
https://github.com/OCA/bank-payment/blob/12.0/account_payment_order
or in Shopinvader
https://github.com/shopinvader/odoo-shopinvader/blob/12.0/shopinvader/models/sale.py
While they are deprecated in v13+ https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#tasks-to-do-in-the-migration
So not sure why @bealdav asked to remove them. But I think it doesn't hurt.

bealdav commented:
It's not deprecated in v13, it fails in v13 and serve to nothing in v10, v11, v12 (default behavior) 

https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#tasks-to-do-in-the-migration

It should have been deprecated from v10, but Odoo SA doesn't like deprecate things

hparfr commented:
> "Advanced Attribute Set settings" 

+1

sebastienbeau commented:
> @hparfr @sebastienbeau @bealdav I made most of the changes you asked for in these last commits.
> 
> I've still have some questions scattered in the discussion which I might gather here :
> 
>     * @sebastienbeau should I add a [readonly on attribute's `serialized` field](https://github.com/akretion/odoo-pim/pull/3#discussion_r430443805) when creating a native attribute even if it's already in an invisible group ?

No 

> 
>     * @sebastienbeau should I [add a test to write on attributes](https://github.com/akretion/odoo-pim/pull/3#discussion_r430491264) in attribute_set as I deleted these tests in product_attribute_set ?

yes move the test in the base module

> 
>     * @bealdav should I [remove the `api.multi` from the whole repo](https://github.com/akretion/odoo-pim/pull/3#discussion_r431191859) or it was just to be removed here for the unlink() ?

remove all @api.multi please their are useless

> 
>     * @sebastienbeau @hparfr should I [rename the attribute_set group allowing to relate attributes to other models](https://github.com/akretion/odoo-pim/pull/3#issuecomment-634716897) _"Relate Attribute set's attributes to existing models"_ instead of _"Advanced Attribute Option"_ ?

yes
> 
> 
> Thank you very much ! sparkles

clementmbr commented:
The debate is on fire for the special attribute_set group  hehe :
https://youtu.be/Y1B02ipctRA?t=242

@hparfr voted for "Advanced Attribute Set settings"
@sebastienbeau  for "Relate Attribute set's attributes to existing models"
@bealdav voted neutral

... I'll need more opinions to make a move !

As I've said, IMO "Advanced Attribute Set settings" looks better if we plan to add some features to this group, but I don't what kind of features it can be. Othewise the explicit "Relate Attribute set's attributes to existing models" looks better to me...

clementmbr committed:
[REF] remove @api.multi

clementmbr committed:
[IMP] Add test write on partner's attributes

clementmbr committed:
[IMP] Rename "Advanced Attribute Set settings" group

clementmbr commented:
@rvalyi @sebastienbeau @hparfr @bealdav 
I've made all the requested latest changes. To me it can be merged with the other branch wher we started the modifications.

bealdav commented:
:+1:

sebastienbeau commented:
let's merge it
